helix-core/src/textobject.rs

```diff
@@ -45,6 +49,54 @@ fn find_word_boundary(slice: RopeSlice, mut pos: usize, direction: Direction) ->
     pos
 }
 
+fn find_paragraph_boundary(slice: RopeSlice, mut pos: usize, direction: Direction) -> usize {
+    // if pos is at non-empty line ending or when going forward move one character left
+    if (!char_is_line_ending(slice.char(pos.saturating_sub(1).min(slice.len_chars())))
```

Comment:
Regarding adding/subtracting from `pos`: keep in mind that CRLF is a possible line ending as well, and is actually two characters.  I don't know if that actually matters here, but I'm wondering if this code might do weird things in documents with CRLF line endings.  We might need to replace some of these with calls to `helix_core::graphemes::prev/next_grapheme_boundary()`.

(In fact, as a general rule, it's best to work in terms of graphemes first, and then switch to char index addition/subtraction only as an optimization in places where we can prove it's correct.  Direct manipulation of char indices is *usually* incorrect.)

Reply:
I believe both `\r` and `\n` is considered newlines right? Here I just want to check the previous character, it doesn't really matter what kind of line ending is it as long as it's line ending.

Reply:
I switched some parts to use graphemes already but I am not sure how to switch the iterator below to use grapheme.